pkg/operator/controller/infrastructure/controller.go

```diff
@@ -0,0 +1,99 @@
+	"k8s.io/client-go/tools/record"
+
+	configv1 "github.com/openshift/api/config/v1"
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	runtimecontroller "sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "infrastructure_controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+func New(mgr manager.Manager) (runtimecontroller.Controller, error) {
+	reconciler := &reconciler{
+		client:   mgr.GetClient(),
+		cache:    mgr.GetCache(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := runtimecontroller.New(controllerName, mgr, runtimecontroller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &configv1.Infrastructure{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type reconciler struct {
+	client   client.Client
+	cache    cache.Cache
+	recorder record.EventRecorder
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	infraConfig := &configv1.Infrastructure{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infraConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get infrastructure 'cluster': %v", err)
+	}
+
+	if infraConfig.Status.PlatformStatus.Type != configv1.AWSPlatformType {
+		log.Info("skipping reconciliation for infrastructure cluster; platform is not %s", "type", configv1.AWSPlatformType)
+		return reconcile.Result{}, nil
+	}
+
+	if !awsPlatformStatusesEqual(*r.config.Status.PlatformStatus.AWS, *infraConfig.Status.PlatformStatus.AWS) {
+		// The infrastructure is safe to process, so ensure it.
+		if err := r.ensureDNSClients(); err != nil {
+			switch e := err.(type) {
+			case retryable.Error:
+				log.Error(e, "got retryable error; requeueing", "after", e.After())
+				return reconcile.Result{RequeueAfter: e.After()}, nil
+			default:
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+
+	return reconcile.Result{}, nil
+}
+
+// awsPlatformStatusesEqual compares two AWSPlatformStatus values.
+// Returns true if the provided values should be considered equal
+// for the purpose of determining whether an update is necessary,
+// false otherwise.
+func awsPlatformStatusesEqual(a, b configv1.AWSPlatformStatus) bool {
+	conditionCmpOpts := []cmp.Option{
+		cmpopts.EquateEmpty(),
+		cmpopts.SortSlices(func(a, b configv1.AWSServiceEndpoint) bool { return a.Name < a.Name || a.URL < b.URL }),
+	}
+	if !cmp.Equal(a, b, conditionCmpOpts...) {
+		return false
+	}
+	return true
+}
+
+// ensureDNSClients...
+func (r *reconciler) ensureDNSClients() error {
+	// TODO
```

Comment:
The DNS controller holds a reference to the provider object, and that reference needs to be updated, so my thinking was that the DNS controller should watch the cluster config and handle updates.  What do you think of that approach?  Proof of concept:  https://github.com/openshift/cluster-ingress-operator/pull/417